### PR TITLE
Ensure a prescribed threading layer can load in CI.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title

### CODE REVIEW

File: buildscripts/incremental/test.sh [hunk 1]
```diff
@@ -41,7 +41,34 @@ else
     esac
 fi
 
+# If TEST_THREADING is set in the env, then check that Numba agrees that the
+# environment can support the requested threading.
+function check_sysinfo() {
+    cmd="import os;\
+         from numba.misc.numba_sysinfo import get_sysinfo;\
+         print(get_sysinfo()['$1 Threading'])"
+    python -c "$cmd"
```
Comment (stuartarchibald) [inline buildscripts/incremental/test.sh:50]: Would an `eval` on the result of the look and followed by an `assert` in the string command be as clear as the additional "check result of query" in bash? Would save having to store the print into a bash var and test it later.

Comment (sklam) [reply buildscripts/incremental/test.sh:50]: an uncaught exception should cause the python process to exit `1`. Or you can also call `os.exit(1)`.

File: buildscripts/incremental/test.sh [hunk 2]
```diff
@@ -41,7 +41,34 @@ else
...
+}
 
+if [[ "$TEST_THREADING" ]]; then
+    if [[ "$TEST_THREADING" == "tbb" ]]; then
+        tstat=$(check_sysinfo "TBB")
```
Comment (sklam) [inline buildscripts/incremental/test.sh:55]: I find it hard to tell if this is running or not. Can use more `echo` when it passes

**Review by sklam — COMMENTED** [review]
Thanks for the patch. I have a few comments.

**Comment by sklam** [comment]
Shouldn't this fail w/o #7605?

File: buildscripts/incremental/test.sh (hunk 1 above)
Comment (stuartarchibald) [reply buildscripts/incremental/test.sh:50]: Indeed, I'll give this a try to limit mixing two languages.

**Comment by stuartarchibald** [comment]
> Shouldn't this fail w/o #7605?

I'm not entirely sure but I think not, #7605 adds both the check for the TBB version and also the correction to the library name. If this branch gains just the check for the TBB version I think it would fail, I'll try that now.

**Comment by stuartarchibald** [comment]
bcf1dc2 forces the TBB version check

**Comment by stuartarchibald** [comment]
Seemingly unrelated fail in linux, python 3.7, np1.18 with coverage:
```
2021-11-30T10:31:30.0079662Z INFO: Running slice of discovered tests: (3,None,19)
```

```
2021-11-30T10:35:40.6235501Z ======================================================================
2021-11-30T10:35:40.6236173Z ERROR: test_linspace_2 (numba.tests.test_dyn_array.TestLinspace)
2021-11-30T10:35:40.6237729Z ----------------------------------------------------------------------
2021-11-30T10:35:40.6238362Z Traceback (most recent call last):
2021-11-30T10:35:40.6238960Z   File "/home/vsts/work/1/s/numba/core/typeinfer.py", line 155, in propagate
2021-11-30T10:35:40.6239521Z     constraint(typeinfer)
2021-11-30T10:35:40.6240108Z   File "/home/vsts/work/1/s/numba/core/typeinfer.py", line 578, in __call__
2021-11-30T10:35:40.6240697Z     self.resolve(typeinfer, typevars, fnty)
2021-11-30T10:35:40.6241288Z   File "/home/vsts/work/1/s/numba/core/typeinfer.py", line 601, in resolve
2021-11-30T10:35:40.6241955Z     sig = typeinfer.resolve_call(fnty, pos_args, kw_args)
2021-11-30T10:35:40.6242705Z   File "/home/vsts/work/1/s/numba/core/typeinfer.py", line 1555, in resolve_call
2021-11-30T10:35:40.6243357Z     return self.context.resolve_function_type(fnty, pos_args, kw_args)
2021-11-30T10:35:40.6244009Z   File "/home/vsts/work/1/s/numba/core/typing/context.py", line 213, in resolve_function_type
2021-11-30T10:35:40.6244603Z     raise last_exception
2021-11-30T10:35:40.6245174Z   File "/home/vsts/work/1/s/numba/core/typing/context.py", line 196, in resolve_function_type
2021-11-30T10:35:40.6245830Z     res = self._resolve_user_function_type(func, args, kws)
2021-11-30T10:35:40.6246449Z   File "/home/vsts/work/1/s/numba/core/typing/context.py", line 248, in _resolve_user_function_type
2021-11-30T10:35:40.6247431Z     return func.get_call_type(self, args, kws)
2021-11-30T10:35:40.6248051Z   File "/home/vsts/work/1/s/numba/core/types/functions.py", line 331, in get_call_type
2021-11-30T10:35:40.6248630Z     failures.raise_error()
2021-11-30T10:35:40.6249207Z   File "/home/vsts/work/1/s/numba/core/types/functions.py", line 227, in raise_error
2021-11-30T10:35:40.6249808Z     raise errors.TypingError(self.format())
2021-11-30T10:35:40.6250507Z numba.core.errors.TypingError: No implementation of function Function(<function linspace at 0x7f07cf2e6f80>) found for signature:
2021-11-30T10:35:40.6251169Z  
2021-11-30T10:35:40.6251631Z  >>> linspace(int64, int64)
2021-11-30T10:35:40.6252078Z  
2021-11-30T10:35:40.6252575Z There are 4 candidate implementations:
2021-11-30T10:35:40.6253414Z       - Of which 4 did not match due to:
2021-11-30T10:35:40.6254396Z       Overload in function '_OverloadWrapper._build.<locals>.ol_generated': File: numba/core/overload_glue.py: Line 131.
2021-11-30T10:35:40.6255326Z         With argument(s): '(int64, int64)':
2021-11-30T10:35:40.6255953Z        Rejected as the implementation raised a specific error:
2021-11-30T10:35:40.6256662Z          TypingError: Failed in nopython mode pipeline (step: native lowering)
2021-11-30T10:35:40.6257331Z        Failed in nopython mode pipeline (step: nopython frontend)
2021-11-30T10:35:40.6258038Z        No implementation of function Function(<function linspace at 0x7f07cf2e6f80>) found for signature:
2021-11-30T10:35:40.6258662Z         
2021-11-30T10:35:40.6259155Z         >>> linspace(int64, int64, Literal[int](50))
2021-11-30T10:35:40.6259662Z         
2021-11-30T10:35:40.6260197Z        There are 4 candidate implementations:
2021-11-30T10:35:40.6261006Z              - Of which 4 did not match due to:
2021-11-30T10:35:40.6268797Z              Overload in function '_OverloadWrapper._build.<locals>.ol_generated': File: numba/core/overload_glue.py: Line 131.
2021-11-30T10:35:40.6269952Z                With argument(s): '(int64, int64, int64)':
2021-11-30T10:35:40.6270507Z               Rejected as the implementation raised a specific error:
2021-11-30T10:35:40.6271306Z                 TypingError: Failed in nopython mode pipeline (step: native lowering)
2021-11-30T10:35:40.6271877Z               Failed in nopython mode pipeline (step: nopython frontend)
2021-11-30T10:35:40.6272679Z               No conversion from array(float64, 1d, C) to array(float64, 1d, C) for '$40return_value.1', defined
```

**Comment by stuartarchibald** [comment]
I can reproduce the above locally:
```
conda create -n _tmp_debug7600 -c numba python=3.7 numpy=1.18 gcc_linux-64 gxx_linux-64 gitpython scipy=1.6.2 llvmlite=0.38.0rc1 ipython jinja2 pygments pyyaml
```
Then install Numba merging this branch with mainline. Then do:
```
./runtests.py -b -j 3,None,19 --exclude-tags='long_running' -m 2 -- numba.tests
```
the hit rate for the error is unfortunately pretty low.

**Comment by stuartarchibald** [comment]
Cannot reproduce the issue on mailine (in 20 runs), I'd guess the problem on this branch is caused by some state from loading `numba.np.ufunc.parallel`, as that's the only material change to the python code in this patch.

File: buildscripts/incremental/test.sh (hunk 1 above)
Comment (sklam) [reply buildscripts/incremental/test.sh:50]: @stuartarchibald, do you want to make this change?

Comment (stuartarchibald) [reply buildscripts/incremental/test.sh]: Done in 1b4666b. Now uses `assert`.

File: buildscripts/incremental/test.sh (hunk 2 above)
Comment (stuartarchibald) [reply buildscripts/incremental/test.sh]: I've updated the python snippet to print on success in 1b4666b

File: buildscripts/incremental/test.sh [hunk 3]
```diff
@@ -41,7 +41,31 @@ else
     esac
 fi
 
+# If TEST_THREADING is set in the env, then check that Numba agrees that the
+# environment can support the requested threading.
+function check_sysinfo() {
+    cmd="import os;\
+         from numba.misc.numba_sysinfo import get_sysinfo;\
+         assert get_sysinfo()['$1 Threading'] is True, 'Threading layer $1 '\
```
Comment (sklam) [inline buildscripts/incremental/test.sh:49]: Does this really need to be an identity check?
```suggestion
         assert get_sysinfo()['$1 Threading'], 'Threading layer $1 '\
```

**Review by sklam — COMMENTED** [review]
There's now a conflict and a small comment:

File: buildscripts/incremental/test.sh (hunk 3 above)
Comment (stuartarchibald) [reply buildscripts/incremental/test.sh:49]: I thought about this when writing it, I don't think there's any enforced return type for the output of `get_sysinfo` as it's largely informational only. My concern was that someone might patch this output to return a string like "It's working" or "It's not working", at which point a boolean assert would alias. If you feel differently am happy to change.

**Review by sklam — APPROVED** [review]
Thanks for the patch!

File: buildscripts/incremental/test.sh (hunk 3 above)
Comment (sklam) [reply buildscripts/incremental/test.sh:49]: fair. we can leave it as is

**Pull request merged**